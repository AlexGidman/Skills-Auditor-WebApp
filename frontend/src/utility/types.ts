export const SYSTEM_ROLES = ["StaffUser", "Manager", "Admin"];
export const STAFF_USER = SYSTEM_ROLES[0];
export const MANAGER_SR = SYSTEM_ROLES[1];
export const ADMIN = SYSTEM_ROLES[2];

export const JOB_ROLES = ["Mid-Level Developer", "Senior Developer", "Manager"];
export const MIDLEVEL_DEVELOPER = JOB_ROLES[0];
export const SENIOR_DEVELOPER = JOB_ROLES[1];
export const MANAGER_JR = JOB_ROLES[2];

export const TOKEN_COOKIE = "token";

// Week in seconds
export const TOKEN_LIFE = 60 * 60 * 24 * 7;

export const SKILL_LEVELS = ["None", "Basic", "Intermediate", "High", "Expert"];
export const NONE = SKILL_LEVELS[0];
export const BASIC = SKILL_LEVELS[1];
export const INTERMEDIATE = SKILL_LEVELS[2];
export const HIGH = SKILL_LEVELS[3];
export const EXPERT = SKILL_LEVELS[4];

export type User = {
    firstName: string;
    lastName: string;
    email: string;
    password: string;
};
