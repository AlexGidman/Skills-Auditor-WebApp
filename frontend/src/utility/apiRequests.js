import axios from "axios";

const categoryEndpoint = "/api/category";
const skillEndpoint = "/api/skill";
const userEndpoint = "/api/user";
const directReportEndpoint = "/api/directreport";
const loginEndpoint = "/api/login";
const userSkillEndpoint = "/api/userskill";

//CATEGORY
export const getAllCategories = async (authHeaders) => {
    return axios.get(categoryEndpoint, authHeaders);
};

export const getCategory = async (authHeaders, id) => {
    return axios.get(`${categoryEndpoint}/${id}`, authHeaders);
};

export const addCategory = async (authHeaders, name) => {
    const data = {
        name,
    };
    return axios.post(`${categoryEndpoint}`, data, authHeaders);
};

export const updateCategory = async (authHeaders, id, name) => {
    const data = {
        id,
        name,
    };
    return axios.put(categoryEndpoint, data, authHeaders);
};

export const deleteCategory = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(categoryEndpoint, { ...authHeaders, data });
};

//SKILL
export const getAllSkills = async (authHeaders) => {
    return axios.get(skillEndpoint, authHeaders);
};

export const getSkill = async (authHeaders, id) => {
    return axios.get(`${skillEndpoint}/${id}`, authHeaders);
};

export const addSkill = async (authHeaders, name, category_id) => {
    const data = {
        name,
        category_id,
    };
    return axios.post(`${skillEndpoint}`, data, authHeaders);
};

export const updateSkill = async (authHeaders, id, name, categoryId) => {
    const data = {
        id,
        name,
        category_id: categoryId,
    };
    return axios.put(skillEndpoint, data, authHeaders);
};

export const deleteSkill = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(skillEndpoint, { ...authHeaders, data });
};

//USER
export const getUserDetails = async (authHeaders, id) => {
    return axios.get(`${userEndpoint}/${id}`, authHeaders);
};

export const getUserFromToken = async (authHeaders) => {
    return axios.get(`${userEndpoint}/token`, authHeaders);
};

export const updateUserDetails = async (
    authHeaders,
    id,
    forename,
    surname,
    email,
    password,
    job_role,
    system_role,
) => {
    const data = {
        id,
        forename,
        surname,
        email,
        password,
        job_role,
        system_role,
    };

    return axios.put(`${userEndpoint}`, data, authHeaders);
};

export const addUser = async (
    authHeaders,
    forename,
    surname,
    email,
    password,
    job_role,
    system_role,
) => {
    const data = {
        forename,
        surname,
        email,
        password,
        job_role,
        system_role,
    };
    return axios.post(userEndpoint, data, authHeaders);
};

export const deleteUser = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(userEndpoint, { ...authHeaders, data });
};

//DIRECT REPORTS
export const getAllDirectReports = async (authHeaders, id) => {
    return axios.get(`${directReportEndpoint}/${id}`, authHeaders);
};

export const addDirectReport = async (authHeaders, userId, reportId) => {
    const data = {
        user_id: userId,
        report_id: reportId,
    };
    return axios.post(directReportEndpoint, data, authHeaders);
};

//STAFF SKILLS
export const getAllStaffSkills = async (authHeaders, id) => {
    return axios.get(`${userSkillEndpoint}/byuser/${id}`, authHeaders);
};

export const getStaffSkill = async (authHeaders, id) => {
    return axios.get(`${userSkillEndpoint}/${id}`, authHeaders);
};

export const deleteStaffSkill = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(userSkillEndpoint, { ...authHeaders, data });
};

export const addUserSkill = async (
    authHeaders,
    user_id,
    skill_id,
    skill_level,
    notes,
    expiry_date,
) => {
    const data = {
        user_id,
        skill_id,
        skill_level,
        notes,
        expiry_date,
    };
    return axios.post(`${userSkillEndpoint}`, data, authHeaders);
};

export const updateStaffSkill = async (authHeaders, id, skill_level, notes) => {
    const data = {
        id,
        skill_level,
        notes,
    };

    return axios.put(`${userSkillEndpoint}`, data, authHeaders);
};

//LOGIN

export const postLogin = async (email, password) => {
    const credentials = { email, password };
    return axios.post(`${loginEndpoint}`, credentials);
};

export const validateToken = async (authHeaders) => {
    return axios.get(`${loginEndpoint}/validate`, authHeaders);
};
