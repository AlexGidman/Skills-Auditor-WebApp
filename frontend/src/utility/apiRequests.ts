import axios, { AxiosRequestHeaders } from "axios";

const categoryEndpoint = "/api/category";
const skillEndpoint = "/api/skill";
const userEndpoint = "/api/user";
const directReportEndpoint = "/api/directreport";
const loginEndpoint = "/api/login";
const userSkillEndpoint = "/api/userskill";

//CATEGORY
export const getAllCategories = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(categoryEndpoint, authHeaders);
};

export const getCategory = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${categoryEndpoint}/${id}`, authHeaders);
};

export const addCategory = async (authHeaders: AxiosRequestHeaders, name: string) => {
    const data = {
        name,
    };
    return axios.post(`${categoryEndpoint}`, data, authHeaders);
};

export const updateCategory = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    name: string,
) => {
    const data = {
        id,
        name,
    };
    return axios.put(categoryEndpoint, data, authHeaders);
};

export const deleteCategory = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(categoryEndpoint, { ...authHeaders, data });
};

//SKILL
export const getAllSkills = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(skillEndpoint, authHeaders);
};

export const getSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${skillEndpoint}/${id}`, authHeaders);
};

export const addSkill = async (
    authHeaders: AxiosRequestHeaders,
    name: string,
    categoryId: string,
) => {
    const data = {
        name,
        categoryId,
    };
    return axios.post(`${skillEndpoint}`, data, authHeaders);
};

export const updateSkill = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    name: string,
    categoryId: string,
) => {
    const data = {
        id,
        name,
        categoryId: categoryId,
    };
    return axios.put(skillEndpoint, data, authHeaders);
};

export const deleteSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(skillEndpoint, { ...authHeaders, data });
};

//USER
export const getUserDetails = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userEndpoint}/${id}`, authHeaders);
};

export const getUserFromToken = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(`${userEndpoint}/token`, authHeaders);
};

export const updateUserDetails = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    firstName: string,
    lastName: string,
    email: string,
    password: string,
    jobRole: string, // TODO: change when job role type changes
    systemRole: string, // TODO: change when job system type changes
) => {
    const data = {
        id,
        firstName,
        lastName,
        email,
        password,
        jobRole,
        systemRole,
    };

    return axios.put(`${userEndpoint}`, data, authHeaders);
};

export const addUser = async (
    authHeaders: AxiosRequestHeaders,
    firstName: string,
    lastName: string,
    email: string,
    password: string,
    jobRole: string, // TODO: change when job role type changes
    systemRole: string, // TODO: change when job system type changes
) => {
    const data = {
        firstName,
        lastName,
        email,
        password,
        jobRole,
        systemRole,
    };
    return axios.post(userEndpoint, data, authHeaders);
};

export const deleteUser = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(userEndpoint, { ...authHeaders, data });
};

//DIRECT REPORTS
export const getAllDirectReports = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${directReportEndpoint}/${id}`, authHeaders);
};

export const addDirectReport = async (
    authHeaders: AxiosRequestHeaders,
    userId: string,
    reportId: string,
) => {
    const data = {
        userId: userId,
        reportId: reportId,
    };
    return axios.post(directReportEndpoint, data, authHeaders);
};

//STAFF SKILLS
export const getAllStaffSkills = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userSkillEndpoint}/byuser/${id}`, authHeaders);
};

export const getStaffSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userSkillEndpoint}/${id}`, authHeaders);
};

export const deleteStaffSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(userSkillEndpoint, { ...authHeaders, data });
};

export const addUserSkill = async (
    authHeaders: AxiosRequestHeaders,
    userId: string,
    skillId: string,
    skillLevel: number, // TODO: chnage this when skill level type changes
    notes: string,
    expiryDate: Date,
) => {
    const data = {
        userId,
        skillId,
        skillLevel,
        notes,
        expiryDate,
    };
    return axios.post(`${userSkillEndpoint}`, data, authHeaders);
};

export const updateStaffSkill = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    skillLevel: number, // TODO: chnage this when skill level type changes
    notes: string,
) => {
    const data = {
        id,
        skillLevel,
        notes,
    };

    return axios.put(`${userSkillEndpoint}`, data, authHeaders);
};

//LOGIN

export const postLogin = async (email: string, password: string) => {
    const credentials = { email, password };
    return axios.post(`${loginEndpoint}`, credentials);
};

export const validateToken = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(`${loginEndpoint}/validate`, authHeaders);
};
