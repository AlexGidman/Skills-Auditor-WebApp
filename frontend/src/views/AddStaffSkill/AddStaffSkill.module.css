.button {
    margin: 3%;
    flex: 42%;
    min-width: 14rem;
}

.form {
    display: flex;
    flex-wrap: wrap;
    margin: 0 10%;
}

.item {
    min-width: 10rem;
    margin: 1% 3%;
    flex: 44%;
}

.textItem {
    min-width: 10rem;
    margin: 1% 3%;
    flex: 44%;
}

.dateItem {
    min-width: 10rem;
    margin: 2% 3%;
    flex: 44%;
}

.linkButton {
    box-sizing: border-box;
    margin: 3%;
    flex: 42%;
    min-width: 14rem;
}

@media (max-width: 900px) {
    .form {
        margin: 0;
    }
}
