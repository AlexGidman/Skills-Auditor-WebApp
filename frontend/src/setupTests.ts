// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import * as apiRequests from "./utility/apiRequests";
import * as reactCookie from "react-cookie";

import { MANAGER_JR, ADMIN, User } from "./utility/types";

jest.mock("uuid", () => {
    return {
        ...jest.requireActual("uuid"),
        v4: jest.fn().mockReturnValue("uuid"),
    };
});

jest.mock("react-cookie", () => {
    return {
        ...jest.requireActual("react-cookie"),
        useCookies: jest.fn(),
    };
});

export const mockApiRequests = {
    getAllCategories: jest.spyOn(apiRequests, "getAllCategories"),
    getCategory: jest.spyOn(apiRequests, "getCategory"),
    updateCategory: jest.spyOn(apiRequests, "updateCategory"),
    getUserDetails: jest.spyOn(apiRequests, "getUserDetails"),
    updateUserDetails: jest.spyOn(apiRequests, "updateUserDetails"),
    getSkill: jest.spyOn(apiRequests, "getSkill"),
    updateSkill: jest.spyOn(apiRequests, "updateSkill"),
    getAllSkills: jest.spyOn(apiRequests, "getAllSkills"),
    addUser: jest.spyOn(apiRequests, "addUser"),
    addDirectReport: jest.spyOn(apiRequests, "addDirectReport"),
    getUserFromToken: jest.spyOn(apiRequests, "getUserFromToken"),
    getAllDirectReports: jest.spyOn(apiRequests, "getAllDirectReports"),
    getAllStaffSkills: jest.spyOn(apiRequests, "getAllStaffSkills"),
    deleteUser: jest.spyOn(apiRequests, "deleteUser"),
    getStaffSkill: jest.spyOn(apiRequests, "getStaffSkill"),
    updateStaffSkill: jest.spyOn(apiRequests, "updateStaffSkill"),
    postLogin: jest.spyOn(apiRequests, "postLogin"),
    addSkill: jest.spyOn(apiRequests, "addSkill"),
    addCategory: jest.spyOn(apiRequests, "addCategory"),
    deleteStaffSkill: jest.spyOn(apiRequests, "deleteStaffSkill"),
    addUserSkill: jest.spyOn(apiRequests, "addUserSkill"),
};

export const mockAdminUser = {
    id: "1",
    firstName: "Admin",
    lastName: "User",
    email: "Admin@email.com",
    jobRole: MANAGER_JR,
    systemRole: ADMIN,
};

export const mockError = { response: { data: { error: new Error("Oops") }, status: 400 } };

beforeAll(() => {
    console.log = jest.fn();
    console.error = jest.fn();
});

beforeEach(() => {
    jest.spyOn(reactCookie, "useCookies").mockImplementation(() => [
        { token: "Bearer token" },
        jest.fn(),
        jest.fn(),
    ]);
    // @ts-ignore
    mockApiRequests.getUserFromToken.mockImplementation(() =>
        Promise.resolve({ data: mockAdminUser }),
    );
});

afterEach(() => {
    jest.resetAllMocks();
});
