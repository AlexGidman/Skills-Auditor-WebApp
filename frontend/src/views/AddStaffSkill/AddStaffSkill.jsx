import React, { useState, useEffect } from "react";
import { useOutletContext, useNavigate, useParams } from "react-router-dom";
import { useImmer } from "use-immer";

import { LoadingPlaceholder } from "../../components/LoadingPlaceholder/LoadingPlaceholder";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { Button } from "../../components/Button/Button";
import { TextInput } from "../../components/TextInput/TextInput";
import { TextAreaInput } from "../../components/TextAreaInput/TextAreaInput";
import { Select } from "../../components/Select/Select";
import {
    useAPI,
    getSelectOptionsFromArray,
    convertSkillLevelName,
    isValidString,
} from "../../utility/helper";

import styles from "./AddStaffSkill.module.css";
import { getAllSkills, addUserSkill, getAllDirectReports } from "../../utility/apiRequests";
import { SKILL_LEVELS } from "../../utility/types";

export const AddStaffSkill = () => {
    const { userId } = useParams();
    const navigate = useNavigate();
    const [currentUser, setShowToast] = useOutletContext();

    const {
        data: directReportsData,
        loading: directReportsLoading,
        error: directReportsError,
    } = useAPI(getAllDirectReports, [currentUser.id]);

    useEffect(() => {
        if (userId) {
            if (
                !!directReportsError ||
                (!!directReportsData &&
                    !directReportsData.filter((directReport) => {
                        return directReport.report.id == userId;
                    }))
            ) {
                setShowToast({
                    error: "You are not authorised to edit this users skills",
                });
                navigate("/");
            }
        }
    }, [currentUser.id, directReportsData, directReportsError, navigate, setShowToast, userId]);

    const { data, loading, error } = useAPI(getAllSkills);

    if (loading || directReportsLoading) return <LoadingPlaceholder />;
    if (error) return <ErrorMessage error={error} />;

    const sortedSkills = data.sort((a, b) => a.name.localeCompare(b.name));
    return (
        <>
            <header className={styles.content}>
                <h1>Add Staff Skill</h1>
            </header>
            <section>
                <Form data={sortedSkills} userId={userId ? parseInt(userId) : currentUser.id} />
            </section>
        </>
    );
};

const Form = ({ data, userId }) => {
    const [, setShowToast] = useOutletContext();
    const [skillLevelValue, setskillLevelValue] = useState();
    const [inputValues, setInputValues] = useImmer({
        user_id: userId,
        skill_id: data ? data[0].id : null,
        skill_level: convertSkillLevelName(SKILL_LEVELS[0]),
        notes: "",
        expiry_date: null,
    });

    const navigate = useNavigate();

    const [formChanged, setFormChanged] = useState(false);
    const [navigateBack, setNavigateBack] = useState(false);

    const {
        data: addData,
        error,
        loading,
        callback: executeStaffAddSkill,
    } = useAPI(addUserSkill, Object.values(inputValues), true);

    useEffect(() => {
        setFormChanged(inputValues.notes !== "" || inputValues.expiry_date !== "");
    }, [inputValues, data]);

    useEffect(() => {
        if (navigateBack && !loading) {
            addData && setShowToast({ success: "Successfully added staff skill" });
            error && setShowToast({ error: "Failed to add staff skill" });
            navigate(-1);
        }
    }, [error, loading, navigate, navigateBack, setShowToast, addData]);

    return (
        <form className={styles.form}>
            {!!data && (
                <Select
                    options={getSelectOptionsFromArray(
                        data.map((skill) => {
                            return skill.id;
                        }),
                        data.map((skill) => {
                            return `${skill.name} - ${skill.category.name}`;
                        }),
                    )}
                    labelText="Select a skill"
                    className={styles.item}
                    value={inputValues.skill_id}
                    onChange={(event) => {
                        setInputValues((draft) => {
                            draft.skill_id = event.target.value;
                        });
                    }}
                    required
                />
            )}

            <Select
                options={getSelectOptionsFromArray(SKILL_LEVELS)}
                labelText="Select a level"
                className={styles.item}
                value={skillLevelValue}
                onChange={(event) => {
                    let value = convertSkillLevelName(event.target.value);
                    setskillLevelValue(event.target.value);
                    setInputValues((draft) => {
                        draft.skill_level = value;
                    });
                }}
                required
            />

            <TextAreaInput
                labelText="Notes"
                className={styles.textItem}
                defaultValue={inputValues.notes}
                name="notes"
                onChange={(event) => {
                    setInputValues((draft) => {
                        draft.notes = event.target.value;
                    });
                }}
                required
            />

            <TextInput
                labelText="Expiry date (optional)"
                className={styles.textItem}
                value={inputValues.expiry_date}
                name="expiryDate"
                onChange={(event) => {
                    setInputValues((draft) => {
                        draft.expiry_date = event.target.value;
                    });
                }}
                type="Date"
                role="input"
            />

            <Button
                className={styles.button}
                loading={loading}
                disabled={!formChanged || !isValidString(inputValues.notes, 5, 255)}
                onClick={(e) => {
                    e.preventDefault();
                    executeStaffAddSkill();
                    setNavigateBack(true);
                }}
            >
                Add Staff Skill
            </Button>
            <Button
                className={styles.button}
                onClick={(e) => {
                    e.preventDefault();
                    navigate(-1);
                }}
            >
                Cancel
            </Button>
        </form>
    );
};
