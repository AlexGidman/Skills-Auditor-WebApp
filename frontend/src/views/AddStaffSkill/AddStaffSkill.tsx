import React, { useState, useEffect } from "react";
import { useOutletContext, useNavigate, useParams } from "react-router-dom";
import { useImmer } from "use-immer";

import { LoadingPlaceholder } from "../../components/LoadingPlaceholder/LoadingPlaceholder";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { Button } from "../../components/Button/Button";
import { TextInput } from "../../components/TextInput/TextInput";
import { TextAreaInput } from "../../components/TextAreaInput/TextAreaInput";
import { Select } from "../../components/Select/Select";
import {
    useAPI,
    getSelectOptionsFromArray,
    convertSkillLevelName,
    isValidString,
} from "../../utility/helper";

import styles from "./AddStaffSkill.module.css";
import { getAllSkills, addUserSkill, getAllDirectReports } from "../../utility/apiRequests";
import { DirectReport, Skill, SKILL_LEVELS } from "../../utility/types";

export const AddStaffSkill = () => {
    const { userId } = useParams();
    const navigate = useNavigate();
    // @ts-ignore TODO fix type for AppOutletContext here
    const [currentUser, setShowToast] = useOutletContext();

    const {
        data: directReportsData,
        loading: directReportsLoading,
        error: directReportsError,
    } = useAPI<DirectReport[]>(getAllDirectReports, [currentUser.id]);

    useEffect(() => {
        if (userId) {
            if (
                !!directReportsError ||
                (!!directReportsData &&
                    !directReportsData.filter((directReport) => {
                        return directReport.report.id === userId; // TODO: check this isn't broken by type string vs number
                    }))
            ) {
                setShowToast({
                    error: "You are not authorised to edit this users skills",
                });
                navigate("/");
            }
        }
    }, [currentUser.id, directReportsData, directReportsError, navigate, setShowToast, userId]);

    const { data, loading, error } = useAPI<Skill[]>(getAllSkills);

    if (loading || directReportsLoading) return <LoadingPlaceholder />;
    if (error) return <ErrorMessage error={error} />;

    const sortedSkills = data ? data.sort((a, b) => a.name.localeCompare(b.name)) : null;
    return (
        <>
            <header className={styles.content}>
                <h1>Add Staff Skill</h1>
            </header>
            <section>
                <Form data={sortedSkills} userId={userId ? parseInt(userId) : currentUser.id} />
            </section>
        </>
    );
};

interface FormProps {
    data: Skill[] | null;
    userId: string;
}

const Form = ({ data, userId }: FormProps) => {
    // @ts-ignore TODO fix type for AppOutletContext here
    const [, setShowToast] = useOutletContext();
    const [skillLevelValue, setskillLevelValue] = useState<string>();
    const [inputValues, setInputValues] = useImmer({
        userId: userId,
        skillId: data ? data[0].id : null,
        skillLevel: convertSkillLevelName(SKILL_LEVELS[0]),
        notes: "",
        expiryDate: "",
    });

    const navigate = useNavigate();

    const [formChanged, setFormChanged] = useState(false);
    const [navigateBack, setNavigateBack] = useState(false);

    const {
        data: addData,
        error,
        loading,
        callback: executeStaffAddSkill,
    } = useAPI(addUserSkill, Object.values(inputValues), true);

    useEffect(() => {
        setFormChanged(inputValues.notes !== "" || inputValues.expiryDate !== "");
    }, [inputValues, data]);

    useEffect(() => {
        if (navigateBack && !loading) {
            addData && setShowToast({ success: "Successfully added staff skill" });
            error && setShowToast({ error: "Failed to add staff skill" });
            navigate(-1);
        }
    }, [error, loading, navigate, navigateBack, setShowToast, addData]);

    return (
        <form className={styles.form}>
            {!!data && (
                <Select
                    options={getSelectOptionsFromArray(
                        data.map((skill) => {
                            return skill.id;
                        }),
                        data.map((skill) => {
                            return `${skill.name} - ${skill.category.name}`;
                        }),
                    )}
                    labelText="Select a skill"
                    className={styles.item}
                    value={inputValues.skillId || undefined}
                    onChange={(event) => {
                        setInputValues((draft) => {
                            draft.skillId = event.target.value;
                        });
                    }}
                    required
                />
            )}
            <Select
                options={getSelectOptionsFromArray(SKILL_LEVELS)}
                labelText="Select a level"
                className={styles.item}
                value={skillLevelValue}
                onChange={(event) => {
                    let value = convertSkillLevelName(event.target.value);
                    setskillLevelValue(event.target.value);
                    setInputValues((draft) => {
                        draft.skillLevel = value;
                    });
                }}
                required
            />
            <TextAreaInput
                labelText="Notes"
                className={styles.textItem}
                defaultValue={inputValues.notes}
                name="notes"
                onChange={(event) => {
                    setInputValues((draft) => {
                        draft.notes = event.target.value;
                    });
                }}
                required
            />
            <TextInput
                labelText="Expiry date (optional)"
                className={styles.textItem}
                value={inputValues.expiryDate}
                name="expiryDate"
                onChange={(event) => {
                    setInputValues((draft) => {
                        draft.expiryDate = event.target.value;
                    });
                }}
                type="Date"
                role="input"
            />
            <Button
                className={styles.button}
                loading={loading}
                disabled={!formChanged || !isValidString(inputValues.notes, 5, 255)}
                onClick={(e) => {
                    e.preventDefault();
                    executeStaffAddSkill();
                    setNavigateBack(true);
                }}
            >
                Add Staff Skill
            </Button>
            <Button
                className={styles.button}
                onClick={(e) => {
                    e.preventDefault();
                    navigate(-1);
                }}
            >
                Cancel
            </Button>
        </form>
    );
};
