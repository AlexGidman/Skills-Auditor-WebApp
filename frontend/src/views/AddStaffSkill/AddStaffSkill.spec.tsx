import { render, screen } from "@testing-library/react";
import { AddStaffSkill } from "./AddStaffSkill";
import { mockAdminUser, mockApiRequests, mockApiResponse, mockError } from "../../setupTests";
import * as ReactDom from "react-router-dom";
import React from "react";
import { Category, Skill } from "../../utility/types";

jest.mock("react-router-dom", () => {
    return {
        ...jest.requireActual("react-router-dom"),
        useOutletContext: jest.fn(),
    };
});

const mockCategory: Category = { id: "1", name: "testname1" };

const mockSkill1: Skill = {
    id: "1",
    name: "Skill1",
    category: mockCategory,
    categoryId: mockCategory.id,
};

const skillLevel = ["None", "Basic", "Intermediate", "High", "Expert"];

describe("AddStaffSkill", () => {
    beforeEach(() => {
        mockApiRequests.getAllDirectReports.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: [] }),
        );
    });
    it("renders correctly when data successful", async () => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue({
            currentUser: mockAdminUser,
            setShowToast: jest.fn(),
        });

        mockApiRequests.getAllSkills.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: [mockSkill1] }),
        );
        const { container } = render(
            <ReactDom.MemoryRouter>
                <AddStaffSkill />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "Add Staff Skill" })).toBeInTheDocument();

        expect(screen.getByRole("combobox", { name: "Select a skill" })).toBeInTheDocument();
        expect(screen.getByRole("combobox", { name: "Select a skill" })).toHaveValue(mockSkill1.id);

        expect(screen.getByRole("combobox", { name: "Select a level" })).toBeInTheDocument();
        expect(screen.getByRole("combobox", { name: "Select a level" })).toHaveValue(skillLevel[0]);

        expect(screen.getByRole("textbox", { name: "Notes:" })).toBeInTheDocument();
        expect(screen.getByRole("textbox", { name: "Notes:" })).toHaveValue("");
        expect(screen.getByRole("input", { name: "Expiry date (optional):" })).toBeInTheDocument();
        expect(screen.getByRole("input", { name: "Expiry date (optional):" })).toHaveValue("");

        expect(screen.getByRole("button", { name: "Add Staff Skill" })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: "Add Staff Skill" })).toBeDisabled();
        expect(screen.getByRole("button", { name: "Cancel" })).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it("renders correctly when data retrieval fails", async () => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue({
            currentUser: mockAdminUser,
            setShowToast: jest.fn(),
        });
        mockApiRequests.getAllSkills.mockImplementation(() => Promise.reject(mockError));
        const { container } = render(
            <ReactDom.MemoryRouter>
                <AddStaffSkill />
            </ReactDom.MemoryRouter>,
        );
        expect(
            await screen.findByRole("heading", { name: "Something went wrong..." }),
        ).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });
});
