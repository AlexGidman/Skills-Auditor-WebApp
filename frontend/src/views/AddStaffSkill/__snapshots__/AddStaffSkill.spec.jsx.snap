// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddStaffSkill renders correctly when data retrieval fails 1`] = `
<div>
  <div
    aria-label="Oops"
    class="component"
    role="alert"
  >
    <h1>
      Something went wrong...
    </h1>
    <p>
      Oops
    </p>
  </div>
</div>
`;

exports[`AddStaffSkill renders correctly when data successful 1`] = `
<div>
  <header
    class="content"
  >
    <h1>
      Add Staff Skill
    </h1>
  </header>
  <section>
    <form
      class="form"
    >
      <div
        class="component item"
      >
        <label
          class="label"
          for="Select-a skill-undefined"
        >
          Select a skill
          :
        </label>
        <select
          aria-label="Select a skill"
          class="select"
          name="Select a skill"
          required=""
        >
          <option
            value="1"
          >
            Skill1 - testname1
          </option>
        </select>
      </div>
      <div
        class="component item"
      >
        <label
          class="label"
          for="Select-a level-undefined"
        >
          Select a level
          :
        </label>
        <select
          aria-label="Select a level"
          class="select"
          name="Select a level"
          required=""
        >
          <option
            value="None"
          >
            None
          </option>
          <option
            value="Basic"
          >
            Basic
          </option>
          <option
            value="Intermediate"
          >
            Intermediate
          </option>
          <option
            value="High"
          >
            High
          </option>
          <option
            value="Expert"
          >
            Expert
          </option>
        </select>
      </div>
      <div
        class="component textItem"
      >
        <label
          class="label"
          for="Notes-undefined"
        >
          Notes
          :
        </label>
        <textarea
          class="input"
          id="Notes-undefined"
          maxlength="100"
          minlength="0"
          name="notes"
          pattern="[a-zA-ZÀ-ȕ-]+$"
          required=""
          rows="4"
        />
      </div>
      <div
        class="component textItem"
      >
        <label
          class="label"
          for="Expiry-date (optional)-undefined"
        >
          Expiry date (optional)
          :
        </label>
        <input
          class="input"
          id="Expiry-date (optional)-undefined"
          maxlength="100"
          minlength="0"
          name="expiryDate"
          pattern="[a-zA-ZÀ-ȕ-]+$"
          role="input"
          type="Date"
          value=""
        />
      </div>
      <button
        class="component button"
        disabled=""
      >
        Add Staff Skill
      </button>
      <button
        class="component button"
      >
        Cancel
      </button>
    </form>
  </section>
</div>
`;
